### PR TITLE
Fix broken drop impl in SFTP module

### LINKED ISSUE #195: Segmentation Fault when `read_to_end`
I got the SegFault when running the `read_to_end` in the codes below.

```rust
let mut sftp_file :ssh2::File = stocks_file.sftp_open()?;
let mut buffer: Vec<u8> = Vec::new();
use std::io::*;
BufReader::with_capacity(1 << 16, sftp_file).read_to_end(&mut buffer)?;
```

```
Thread 1 "integrations_cd" received signal SIGSEGV, Segmentation fault.
sftp_read (buffer_size=65536, buffer=0x555555e289c0 "", handle=0x555555d93360) at libssh2/src/sftp.c:1342
1342        LIBSSH2_SESSION *session = channel->session;
(gdb) bt
#0  sftp_read (buffer_size=65536, buffer=0x555555e289c0 "", handle=0x555555d93360) at libssh2/src/sftp.c:1342
#1  libssh2_sftp_read (hnd=0x555555d93360, buffer=0x555555e289c0 "", buffer_maxlen=65536) at libssh2/src/sftp.c:1738
#2  0x000055555568f0db in <ssh2::sftp::File as std::io::Read>::read () at /home/kunhao/.cargo/registry/src/github.com-1ecc6299db9ec823/ssh2-0.8.0/src/sftp.rs:578
#3  0x000055555561110b in <std::io::buffered::BufReader<R> as std::io::BufRead>::fill_buf () at /home/kunhao/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/buffered.rs:314
#4  <std::io::buffered::BufReader<R> as std::io::Read>::read () at /home/kunhao/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/buffered.rs:274
#5  0x0000555555636c2c in std::io::read_to_end_with_reservation () at /home/kunhao/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/mod.rs:396
#6  std::io::read_to_end () at /home/kunhao/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/mod.rs:363
#7  0x000055555560dea1 in std::io::Read::read_to_end () at /home/kunhao/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/mod.rs:680
```

### PR DESCRIPTION
**Pull request by Ten0**
Resolves #195 (use after free, double free, segfault...)

### CODE REVIEW

**Comment by Ten0** [comment]
Hello, this lib has 12,000 downloads monthly and nobody cares it segfaults on the simplest `let file = session.sftp().open(filename)` ?

**Comment by crisidev** [comment]
I think everyone cares, it is just a about having time in this cahotic life :) I have experienced this behavior as well and I am using scp, which is a poor choice compared to SFTP and does not give the same possibilities. @wez is it possible to prioritize the review of this bugfix?

**Comment by wez** [comment]
Sorry; I haven't had much bandwidth recently. I can't make any promises about when I will have bandwidth.  This PR is complex enough that I don't feel good just eyeballing and merging it based on the CI being successful.
I do appreciate the time you've put into contributing this PR!

**Comment by crisidev** [comment]
I have some time to give this PR a close look if this can help. I will do a review. I don't know if it would be ok, but I would like to help with this create if possible as I am a heavy user of it.

File: src/sftp.rs [hunk 1]
```diff
@@ -400,11 +401,15 @@ impl Sftp {
 
     fn lock(&self) -> Result<LockedSftp, Error> {
         match self.inner.as_ref() {
-            Some(inner) => {
-                let sess = inner.sess.lock();
+            Some(sftp_inner_drop_wrapper) => {
+                let sftp_inner = sftp_inner_drop_wrapper
+                    .0
+                    .as_ref()
+                    .expect("Never unset until shutdown, in which case inner is also unset");
```
Comment (crisidev) [inline src/sftp.rs:408]: Since this method already return a Result, can't we raise an Error here instead of panicing?

Comment (crisidev) [reply src/sftp.rs:408]: On a second thought, this is really a unlikely scenario where a LOT of things went wrong, hence panicing looks to be the right thing to do.

**Review by crisidev — APPROVED** [review]
LGTM.. It would be nice to add some tests if possible.

File: src/sftp.rs (hunk 1 above)
Comment (Ten0) [reply src/sftp.rs:408]: Yes, more than an unlikely scenario, it should not be possible for any user to make that panic trigger, and if it does we want to get a report here => this is what panic is for, as explained by https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#cases-in-which-you-have-more-information-than-the-compiler

**Comment by Ten0** [comment]
>  This PR is complex enough that I don't feel good just eyeballing and merging it based on the CI being successful.

> LGTM.. It would be nice to add some tests if possible.

In case that can perhaps reassure you a bit, we've been using this fork in diverse ways in several places over our codebase for over a month now and haven't had any additional issues since.

**Comment by crisidev** [comment]
This is reassuring and for me the tests are not a blocker for this PR. The issue of not having real tests is that this behaviour (use after-free, segfault, etc) could have been caught by some testing here I think: https://github.com/alexcrichton/ssh2-rs/blob/master/tests/all/sftp.rs#L12

**Comment by Ten0** [comment]
>  This PR is complex enough that I don't feel good just eyeballing and merging it based on the CI being successful.

That seems like good practice when you have something that works and someone comes up with a new feature, but doesn't look worth waiting three months when you have something that is recognized to be so broken there may be [RCE](https://en.m.wikipedia.org/wiki/Arbitrary_code_execution)s in it and someone comes up with a fix.

**Comment by crisidev** [comment]
Again, I am available and I have time to help maintaining this project. I use this library daily..

**Comment by Ten0** [comment]
@alexcrichton Would you consider adding another maintainer ?

**Comment by alexcrichton** [comment]
I think it's fine to add folks, but @wez is the primary maintainer nowadays, I  haven't had time for this crate in quite some time unfortunately.

**Comment by crisidev** [comment]
Any news @wez?

**Comment by wez** [comment]
Alright, thanks for the review @crisidev.  This looks good to me also!

**Comment by wez** [comment]
Published in 0.8.3; thanks for your patience!

**Comment by crisidev** [comment]
Thanks @wez.. Any chance we can expand the maintainer list here? One of my work projects also depends on this library and I would like to help if possible.

**Comment by wez** [comment]
I looked but I don't have full admin access to manage maintainers; @alexcrichton, would you mind either giving me access to do that, or granting @crisidev push access?

**Comment by crisidev** [comment]
Thanks a lot! Tomorrow I'll start working on a pull request adding some examples for some more complex things that can be done with this library, such has dropping into netconf :smile:

**Comment by crisidev** [comment]
@alexcrichton sorry to bother you again. Do you mind to give me push access to the repo so I can start working on some of the outstanding issues? wez cannot do it, but he is fine with this :smile:

**Comment by alexcrichton** [comment]
Oops sorry I missed that, added now!

**Comment by alexcrichton** [comment]
Unfortunately since this is a personal repo, not an org repo, GitHub only allows me to invite collaborators, otherwise I'd give admin if I could...

**Comment by crisidev** [comment]
Thanks a lot!

**Pull request merged**